src/librustdoc/html/highlight.rs

```diff
@@ -41,20 +47,47 @@ crate fn render_with_highlighting(
     }
 
     write_header(out, class);
-    write_code(out, &src, edition);
+    write_code(out, &src, edition, file_span_lo, context, root_path);
     write_footer(out, playground_button);
 }
 
 fn write_header(out: &mut Buffer, class: Option<&str>) {
     write!(out, "<div class=\"example-wrap\"><pre class=\"rust {}\">\n", class.unwrap_or_default());
 }
 
-fn write_code(out: &mut Buffer, src: &str, edition: Edition) {
+/// Convert the given `src` source code into HTML by adding classes for highlighting.
+///
+/// This code is used to render code blocks (in the documentation) as well as the source code pages.
+///
+/// Some explanations on the last arguments:
+///
+/// In case we are rendering a code block and not a source code file, `file_span_lo` value doesn't
+/// matter and `context` will be `None`. To put it more simply: if `context` is `None`, the code
+/// won't try to generate links to an ident definition.
```

Comment:
Can you merge them into the same option so it enforces they must be used together? Then you won't have to document it. Like this:
```rust
fn write_code(
    out: &mut Buffer,
    src: &str,
    edition: Edition,
    context: Option<(&Context<'_>, u32)>,
    root_path: &str,
) {

Reply:
It might be worth it to wrap them into a type directly. I'll check that.

Reply:
I moved the new arguments (including `root_path`) into a new struct called `ContextInfo`.